Test/TestFramework/Shared/TestEventListener.cs

```diff
@@ -10,24 +11,105 @@
 #if NET40
     using Microsoft.Diagnostics.Tracing;
 #endif
+#if NET40 || NET45
+    using System.Runtime.Remoting.Messaging;
+#endif
 
     internal class TestEventListener : EventListener
     {
+#if NET40 || NET45
+
+        public static class CurrentContextEvents
+        {
+            internal const string InternalOperationsMonitorSlotName = "Microsoft.ApplicationInsights.TestEventListener";
+
+            private static Object syncObj = new object();
+
+            public static bool IsEntered()
+            {
+                object data = null;
+                try
+                {
+                    data = CallContext.LogicalGetData(InternalOperationsMonitorSlotName);
+                }
+                catch (Exception)
+                {
+                    // CallContext may fail in partially trusted environment
+                }
+
+                return data != null;
+            }
+
+            public static void Enter()
+            {
+                try
+                {
+                    CallContext.LogicalSetData(InternalOperationsMonitorSlotName, syncObj);
+                }
+                catch (Exception)
+                {
+                    // CallContext may fail in partially trusted environment
```

Comment:
Does it mean that some tests may pass in this case? No exception will happen and it looks like no traces will be collected (including error traces) because enter() failed for test trace listener itself.